physics/hierarchical_system_body.hpp

```diff
@@ -0,0 +1,151 @@
+                                                 LocalFrameTag::kFrameTag,
+                                                 /*frame_is_inertial=*/false>;
+        static DegreesOfFreedom<SystemBarycentre> const system_barycentre = {
+            SystemBarycentre::origin, Velocity<SystemBarycentre>()};
+        static Identity<SystemBarycentre, Frame> const id_bf;
+        static Identity<Frame, SystemBarycentre> const id_fb;
+
+        // Jacobi coordinates for |system|, with satellite subsystems treated
+        // as point masses at their barycentres.
+        JacobiCoordinates<Frame> jacobi_coordinates(*system.primary);
+        result.bodies.emplace_back(std::move(system.primary));
+
+        // The |n|th element of |satellite_degrees_of_freedom| contains the list
+        // of degrees of freedom of the bodies in the |n|th satellite subsystem
+        // with respect to their barycentre.
+        std::vector<std::vector<RelativeDegreesOfFreedom<Frame>>>
+            satellite_degrees_of_freedom;
+
+        // Fill |satellite_degrees_of_freedom|, |jacobi_coordinates|, and
+        // |result.bodies|.
+        for (auto const& subsystem : system.satellites) {
+          BarycentricSubystem barycentric_satellite_subsystem =
+              to_barycentric(*subsystem);
+          satellite_degrees_of_freedom.emplace_back(std::move(
+              barycentric_satellite_subsystem.barycentric_degrees_of_freedom));
+          jacobi_coordinates.Add(
+              *barycentric_satellite_subsystem.equivalent_body,
+              subsystem->jacobi_osculating_elements);
+          std::move(barycentric_satellite_subsystem.bodies.begin(),
+                    barycentric_satellite_subsystem.bodies.end(),
+                    std::back_inserter(result.bodies));
+        }
+
+        std::vector<DegreesOfFreedom<SystemBarycentre>>
+            barycentres_of_subsystems;
+        {
+          // TODO(egg): should BarycentricDegreesOfFreedom return
+          // DegreesOfFreedom instead of RelativeDegreesOfFreedom?  In what
+          // frame?
+          auto const barycentric_dof =
+              jacobi_coordinates.BarycentricDegreesOfFreedom();
+          for (auto const& dof : barycentric_dof) {
+            barycentres_of_subsystems.push_back(system_barycentre + id_fb(dof));
+          }
+        }
+
+        // Fill |result.barycentric_degrees_of_freedom|.
+        // The primary.
+        result.barycentric_degrees_of_freedom.emplace_back(
+            id_bf(barycentres_of_subsystems.front() - system_barycentre));
+        // The bodies in satellite subsystems.
+        for (int n = 0; n < satellite_degrees_of_freedom.size(); ++n) {
+          // |n + 1| because the primary is at |barycentres_of_subsystems[0]|.
+          DegreesOfFreedom<SystemBarycentre> const subsystem_barycentre =
+              barycentres_of_subsystems[n + 1];
+          for (auto const& body_dof_wrt_subsystem_barycentre :
+               satellite_degrees_of_freedom[n]) {
```

Comment:
clang-format says otherwise; that `:` isn't an operator, and `auto const& body_dof_wrt_subsystem_barycentre : satellite_degrees_of_freedom[n]` isn't an expression (or anything really, since the `:` is part of the syntax for the range-based for loop), so let's go with what clang-format does.